dask/array/tests/test_routines.py

```diff
@@ -1286,3 +1286,121 @@ def test_result_type():
+    '...abc,...abcd->...d',
+    'ab...,b->ab...',
+    # https://github.com/dask/dask/pull/3412#discussion_r182413444
+    'aa->a',
+    'ab,ab,c->c',
+    'aab,bc->ac',
+    'aab,bcc->ac',
+    'fdf,cdd,ccd,afe->ae',
+    'fff,fae,bef,def->abd',
+])
+def test_einsum(einsum_signature):
+    # einsum label dimensions
+    dimensions = {'a': 5, 'b': 6, 'c': 7,
+                  'd': 5, 'e': 6, 'f': 10,
+                  'g': 1, 'h': 2, '*': 11}
+
+    # dimension chunks sizes
+    chunks = {'a': (2, 3), 'b': (2, 3, 1), 'c': (2, 3, 2),
+              'd': (4, 1), 'e': (2, 4),    'f': (1, 2, 3, 4),
+              'g': 1,      'h': (1, 1),    '*': 11}
+
+    def shape_from_string(s):
+        return tuple(dimensions[c] for c in s)
+
+    def chunks_from_string(s):
+        return tuple(chunks[c] for c in s)
+
+    input_sigs = (einsum_signature.split('->')[0]
+                                  .replace("...", "*")
+                                  .split(','))
+
+    shapes = [shape_from_string(s) for s in input_sigs]
+    chunks = [chunks_from_string(s) for s in input_sigs]
+
+    np_inputs = [np.random.random(s) for s in shapes]
+    da_inputs = [da.from_array(i, chunks=c) for i, c in zip(np_inputs, chunks)]
+
+    assert_eq(np.einsum(einsum_signature, *np_inputs),
+              da.einsum(einsum_signature, *da_inputs))
+
+
+def test_einsum_broadcasting_contraction():
+    a = np.random.rand(1, 5, 4)
+    b = np.random.rand(4, 6)
+    c = np.random.rand(5, 6)
+    d = np.random.rand(10)
+
+    d_a = da.from_array(a, chunks=(1,(2,3),(2,2)))
+    d_b = da.from_array(b, chunks=((2,2), (4,2)))
+    d_c = da.from_array(c, chunks=((2,3), (4,2)))
+    d_d = da.from_array(d, chunks=((7,3)))
+
+    np_res = np.einsum('ijk,kl,jl', a, b, c)
+    da_res = da.einsum('ijk,kl,jl', d_a, d_b, d_c)
+    assert_eq(np_res, da_res)
+
+    np_res = np.einsum('ijk,kl,jl,i->i', a, b, c, d)
+    da_res = da.einsum('ijk,kl,jl,i->i', d_a, d_b, d_c, d_d)
+    assert_eq(np_res, da_res)
+
+
+def test_einsum_broadcasting_contraction2():
+    a = np.random.rand(1, 1, 5, 4)
+    b = np.random.rand(4, 6)
+    c = np.random.rand(5, 6)
+    d = np.random.rand(7, 7)
+
+    d_a = da.from_array(a, chunks=(1, 1,(2,3),(2,2)))
+    d_b = da.from_array(b, chunks=((2,2), (4,2)))
+    d_c = da.from_array(c, chunks=((2,3), (4,2)))
+    d_d = da.from_array(d, chunks=((7,3)))
+
+    np_res = np.einsum('abjk,kl,jl', a, b, c)
+    da_res = da.einsum('abjk,kl,jl', d_a, d_b, d_c)
+    assert_eq(np_res, da_res)
+
+    np_res = np.einsum('abjk,kl,jl,ab->ab', a, b, c, d)
+    da_res = da.einsum('abjk,kl,jl,ab->ab', d_a, d_b, d_c, d_d)
+    assert_eq(np_res, da_res)
+
+
+def test_einsum_broadcasting_contraction3():
+    a = np.random.rand(1, 5, 4)
+    b = np.random.rand(4, 1, 6)
+    c = np.random.rand(5, 6)
+    d = np.random.rand(7, 7)
+
+    d_a = da.from_array(a, chunks=(1,(2,3),(2,2)))
+    d_b = da.from_array(b, chunks=((2,2), 1,  (4,2)))
```

Comment:
I'm somewhat surprised that flake8 isn't yelling at you here.  Typically it requires exactly one space after every comma.

Reply:
Fixed